base/abstractdict.jl

```diff
@@ -335,6 +335,75 @@ function _typeddict(d::AbstractDict, others::AbstractDict...)
     Dict{K,V}(d)
 end
 
+function reduce(::typeof(merge), dicts::AbstractVector{T}) where T <: AbstractDict{K,V} where {K,V}
+    isempty(dicts) && throw(ArgumentError("Cannot merge empty array"))
+    return _merge!(empty(first(dicts)),dicts) # use `empty`. `similar` makes routine 20x slower
+end
+
+function reduce(::typeof(merge), d::T, dicts::AbstractVector{T}) where T <: AbstractDict{K,V} where {K,V}
+    isempty(dicts) && return copy(d)
+    return _merge!(empty(first(dicts)),d,dicts) # There is no evidence that calling _merge! twice is less efficient.
+end
+
+# This catches typeof(T) == UnionAll.
+# The conversion of `dicts` to a vector of type Any makes this method much faster.
+function reduce(::typeof(merge), dicts::AbstractVector{T}) where T <: AbstractDict
+    isempty(dicts) && throw(ArgumentError("Cannot merge empty array"))
+    return reduce(merge, Any[x for x in dicts])
+end
+
+function reduce(::typeof(merge), d::AbstractDict, dicts::AbstractVector{T}) where T <: AbstractDict
+    isempty(dicts) && throw(ArgumentError("Cannot merge empty array"))
+    return reduce(merge, d, Any[x for x in dicts])
+end
+
+function reduce(m::typeof(merge), items::AbstractVector)
+    isempty(items) && throw(ArgumentError("Cannot merge empty vector"))
+    length(items) == 1 && return copy(first(items))
+    i1 = firstindex(items) + 1
+    i2 = lastindex(items)
+    return reduce(m,first(items), view(items,i1:i2))
+end
+
+function reduce(m::typeof(merge), d::AbstractDict, items::AbstractVector)
+    isempty(items) && return copy(d)
+    all(x -> isa(x,Dict), items) || return _reduce_dim(merge, d, items, Colon())
+    dictout = _typeddictA(d,items)
+    _merge!(dictout,items)
+end
+
+function _typeddictA(d::AbstractDict{K,V}, others) where {K,V}
+    k, v = K, V
+    for other in others
+        k = promote_type(k, keytype(other))
+        v = promote_type(v, valtype(other))
+    end
+    return Dict{k,v}(d)
+end
+
+function _merge!(dto::AbstractDict, d, dicts::AbstractArray)
+    _merge!(dto,d)
+    for dict in dicts
+        _merge!(dto,dict)
+    end
+    return dto
+end
+
+function _merge!(dto::AbstractDict, dicts::AbstractArray)
+    for d in dicts
+        _merge!(dto,d)
+    end
+    return dto
+end
+
+function _merge!(dto::AbstractDict,  dfrom::AbstractDict)
```

Comment:
Isn't this just `merge!`?

Reply:
Probably. In any case, it is never called. It should be removed.

Reply:
I don't understand. `merge!` takes splatted arguments, but that's for multiple *dictionaries*, not for multiple *pairs*. Here the method only allows two arguments, so it's equivalent to `merge!`.

Reply:
Neither do I ! :) I wasn't thinking of pairs. I'll state how I see it and see if you or agree or see a mistake.
   1. We are talking about the methods `merge!(d::AbstractDict, others::AbstractDict...)` and `_merge!(dto::AbstractDict,  dfrom::AbstractDict)`.
    2. For `Dict`s `a` and `b`, there is no observable difference between `merge!(a,b)` and `_merge!(a,b)`
    3. For the call `merge!(a,b)`, the argument `b` appears in the body in a `Tuple` of one element. The routine must iterate over this tuple.  In `_merge!(a,b)`, there is no tuple and no corresponding iteration. So the latter may be faster than the former.

I did some benchmarks and `_merge!` is indeed faster. How much faster depends of course on the characteristics of the input dicts.  So you gain a bit of efficiency at the expense of writing another (short) method. One can argue for or against `_merge!(a,b)` by arguing that the efficiency outweighs the code complexity, or vice-versa.

Reply:
A concrete example:
```
julia> print(d1)
Dict(7=>9,9=>5,4=>3,2=>8,3=>5,5=>6,1=>1)
julia> print(d2)
Dict(7=>6,9=>6,2=>3,3=>10,5=>9,8=>10,6=>2)
julia> @btime merge!($(copy(d1)),d2);
  133.953 ns (0 allocations: 0 bytes)

julia> @btime _merge!($(copy(d1)),d2);
  87.529 ns (0 allocations: 0 bytes)
```

Reply:
Not sure whether the performance difference is significant given how short these timings are, but if it makes a significant difference in your concrete use case then it's worth filing a separate issue. If the difference is real a two-argument method could be added to `merge!`.

Reply:
Well, the ratio is significant, and I think that's more important. But, the point is moot. I just wrote a discourse post about the general question. After thinking while composing the questions,  I'm quite comfortable with implementing all your suggestions. I think its the better choice. As you say, maybe further optimizations can be made in the future.